iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/metastore/task/IcebergHouseKeeperService.java

```diff
@@ -0,0 +1,131 @@
+  }
+
+  @Override
+  public void run() {
+    LOG.debug("Running IcebergHouseKeeperService...");
+
+    String catalogName = MetastoreConf.getVar(conf, MetastoreConf.ConfVars.ICEBERG_TABLE_EXPIRY_CATALOG_NAME);
+    String dbPattern = MetastoreConf.getVar(conf, MetastoreConf.ConfVars.ICEBERG_TABLE_EXPIRY_DATABASE_PATTERN);
+    String tablePattern = MetastoreConf.getVar(conf, MetastoreConf.ConfVars.ICEBERG_TABLE_EXPIRY_TABLE_PATTERN);
+
+    try (IMetaStoreClient msc = new HiveMetaStoreClient(conf)) {
+      // TODO: Future improvement – modify TableFetcher to return HMS Table API objects directly,
+      // avoiding the need for subsequent msc.getTable calls to fetch each matched table individually
+      List<TableName> tables = getTableFetcher(msc, catalogName, dbPattern, tablePattern).getTables();
+
+      LOG.debug("{} candidate tables found", tables.size());
+
+      for (TableName table : tables) {
+        expireSnapshotsForTable(getIcebergTable(table, msc));
+      }
+    } catch (Exception e) {
+      LOG.error("Exception while running iceberg expiry service on catalog/db/table: {}/{}/{}", catalogName, dbPattern,
+          tablePattern, e);
+    }
+  }
+
+  @VisibleForTesting
+  TableFetcher getTableFetcher(IMetaStoreClient msc, String catalogName, String dbPattern, String tablePattern) {
+    return new TableFetcher.Builder(msc, catalogName, dbPattern, tablePattern).tableTypes(
+            "EXTERNAL_TABLE")
+        .tableCondition(
+            hive_metastoreConstants.HIVE_FILTER_FIELD_PARAMS + "table_type like \"ICEBERG\" ")
+        .build();
+  }
+
+  private Table getIcebergTable(TableName table, IMetaStoreClient msc) throws TException {
+    GetTableRequest request = new GetTableRequest(table.getDb(), table.getTable());
+    return IcebergTableUtil.getTable(conf, msc.getTable(request));
+  }
+
+  /**
+   * Deletes snapshots of an Iceberg table, using the number of threads defined by the
+   * Hive config HIVE_ICEBERG_EXPIRE_SNAPSHOT_NUMTHREADS.
+   * This is largely equivalent to the HiveIcebergStorageHandler.expireSnapshotWithDefaultParams method.
+   *
+   * @param icebergTable the iceberg Table reference
+   */
+  private void expireSnapshotsForTable(Table icebergTable) {
+    LOG.info("Expire snapshots for: {}", icebergTable);
+    ExpireSnapshots expireSnapshots = icebergTable.expireSnapshots();
+
+    int numThreads = conf.getInt(HiveConf.ConfVars.HIVE_ICEBERG_EXPIRE_SNAPSHOT_NUMTHREADS.varname,
+        HiveConf.ConfVars.HIVE_ICEBERG_EXPIRE_SNAPSHOT_NUMTHREADS.defaultIntVal);
+
+    ExecutorService deleteExecutorService = null;
+    try {
+      if (numThreads > 0) {
+        LOG.info("Executing expire snapshots on iceberg table {} with {} threads", icebergTable.name(), numThreads);
+        deleteExecutorService = HiveIcebergUtil.getDeleteExecutorService(icebergTable.name(), numThreads);
+      }
+      if (deleteExecutorService != null) {
```

Comment:
can it be null?

Reply:
yes, in case numThreads == 0 is explicitly set, ExpireSnapshots is not supplied with an executor service, in which case if uses the default one:
```
  RemoveSnapshots(TableOperations ops) {
...
    this.deleteExecutorService = DEFAULT_DELETE_EXECUTOR_SERVICE;
...
```

Reply:
oh yes, I see now, sorry